### ISSUE TITLE
[bitnami/etcd] Scale from 0 or 1 for replica not working

### ISSUE THREAD

**Issue by xgenvn** [bitnami/charts#10009]
### Name and Version

bitnami/etcd:8.1.0

### What steps will reproduce the bug?

1. Start new fresh etcd cluster
```
helm install etcd-test bitnami/etcd --set auth.rbac.create=false --set replicaCount=3 --set initialClusterState=new
```
2. Wait the etcd installation is finished
3. Scale down to 0
4. Scale up again to 3
5. Cluster is unable to start

Same with senarios when scaling down to 1 an up again to 3.

### What is the expected behavior?

Cluster should continue working normally.

### What do you see instead?

Cluster is unable to start

**Comment by xgenvn** [comment]
Another issue which may be related:
1. With `replicaCount=1`
2. Scale up to 3
3. New pods can not automatically register with the cluster (check through `etcdctl member list`)

**Comment by xgenvn** [comment]
#### More details on the setup of scale down issue:

- K8s: typical k3d cluster
- Helm install values
```yaml
auth:
  rbac:
    create: false

replicaCount: 3

persistence:
  size: 2Gi
  accessModes:
    - ReadWriteOnce
```

```
helm install etcd-test bitnami/etcd --values values-etcd.yaml --namespace demo --create-namespace
```
- Scale command
```
k scale statefulset etcd-test --replicas=1 --namespace=demo
```
- Attached logs
[etcd-test-0.log](https://github.com/bitnami/charts/files/8618240/etcd-test-0.log)

---
#### For scale-up issue
- Fresh start with `replicas=1`
- Scale up command
```
k scale statefulset etcd-test --replicas=3 --namespace=demo
```
- Set a key for testing on `etcd-0` pod (which is the first etcd cluster)
```sh
$ etcdctl put testkey "123"
$ etcdctl get testkey
testkey
123
```
- Get the key in another instance (eg: `etcd-1`
```
$ etcdctl get testkey
# return nothing
```
- Output of `etcdctl member list`
```sh
$ etcdctl member list
1b1d9e37ccb16fd9, started, etcd-test-1, http://etcd-test-1.etcd-test-headless.demo.svc.cluster.local:2380, http://etcd-test-1.etcd-test-headless.demo.svc.cluster.local:2379,http://etcd-test.demo.svc.cluster.local:2379, false
```
- Output of `etcdctl member list` (etcd-0)
```sh
$ etcdctl member list
1b1d9e37ccb16fd9, started, etcd-test-1, http://etcd-test-1.etcd-test-headless.demo.svc.cluster.local:2380, http://etcd-test-1.etcd-test-headless.demo.svc.cluster.local:2379,http://etcd-test.demo.svc.cluster.local:2379, false
```

Is there anything that I'm missing on these? I guess it's basic use-cases. I'm not trying to scale etcd that often, it's just started from the issue when boot up K8s cluster or migrate K8s to higher version.

**Comment by migruiz4** [comment]
Hi @xgenvn, I'm sorry for the late response.

After following your step by step instructions, these were my findings:

**Scale down from 3 -> 1**
Fails because it is not a fresh install, you pods are picking the PVCs from a previous install. From your log file:
```
INFO ==> Generating etcd config file using env variables
INFO ==> Detected data from previous deployments <--
WARN ==> Cluster not responding!
WARN ==> Disaster recovery is disabled, the cluster will try to recover on it's own
INFO ==> ** etcd setup finished! **
INFO ==> ** Starting etcd **
```
I was able to scale down by installing, waiting for nodes to be ready, then scaling down.

**Scale up from 1 -> 3 using kubectl scale** 

In this case, there is an issue where `etcd` will fail to scale up when using `k scale` if it was deployed with a single replica. 

It is because the following variables won't be present in the statefulset if deployed with `replicaCount=1`:
https://github.com/bitnami/charts/blob/0b123f47ebc3215a7f3acddc9a752b1ed5552f83/bitnami/etcd/templates/statefulset.yaml#L185-L196

**Scale up from 1 -> 3 using helm upgrade**

There seems to be some kind of race condition when deploying several new nodes at once, where one of the new nodes will fail with the following error:
```json
{"level":"fatal","ts":"2022-05-18T07:22:33.260Z","caller":"etcdmain/etcd.go:204","msg":"discovery failed","error":"error validating peerURLs {ClusterID:f5d976c97d8355be Members:[&{ID:2cab3882123e560c RaftAttributes:{PeerURLs:[http://etcd-test-1.etcd-test-headless.demo.svc.cluster.local:2380] IsLearner:false} Attributes:{Name: ClientURLs:[]}} &{ID:379714abc1eca7dc RaftAttributes:{PeerURLs:[http://etcd-test-0.etcd-test-headless.demo.svc.cluster.local:2380] IsLearner:false} Attributes:{Name:etcd-test-0 ClientURLs:[http://etcd-test-0.etcd-test-headless.demo.svc.cluster.local:2379 http://etcd-test.demo.svc.cluster.local:2379]}}] RemovedMemberIDs:[]}: member count is unequal","stacktrace":"go.etcd.io/etcd/server/v3/etcdmain.startEtcdOrProxyV2\n\t/go/src/go.etcd.io/etcd/release/etcd/server/etcdmain/etcd.go:204\ngo.etcd.io/etcd/server/v3/etcdmain.Main\n\t/go/src/go.etcd.io/etcd/release/etcd/server/etcdmain/main.go:40\nmain.main\n\t/go/src/go.etcd.io/etcd/release/etcd/server/main.go:32\nruntime.main\n\t/go/gos/go1.16.15/src/runtime/proc.go:225"}
```

The issue is probably related to this https://github.com/etcd-io/etcd/issues/10626 and I think it is caused by the two nodes being deployed at the exact same time the node 0 gets restarted to apply statefulset changes.

To workaround this issue, I tried two different solutions:

- Manually scaling 1 by 1, started with `replicaCount=1`, wait for ready, then `replicaCount=2`, then wait for ready again and finally `replicaCount=3`.
- Setting `podManagementPolicy=OrderedReady`, to avoid pods from starting in parallel and not having to increase replicas manually.

**Comment by day1118** [comment]
Hi @migruiz4 
I was looking into the **Scale up from 1 -> 3 using helm upgrade** case.

I also found that there is a race condition, and starting one at a time does resolve the issue. I even found that adding a short `sleep` command before the `etcd-2` pod starts is enough to fix the issue.

Is this race condition tracked anywhere? Are there any plans to investigate it?
Could it be fixed by using a lock etc?
